crates/ruff/src/rules/pylint/rules/import_outside_top_level.rs

```diff
@@ -0,0 +1,44 @@
+use rustpython_parser::ast::Stmt;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::prelude::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for `import` statements that are not at the top-level of a file.
+///
+/// ## Why is this bad?
+/// According to [Pep 8 - Style Guide for Python Code](https://peps.python.org/pep-0008/#imports),
+/// > Imports are always put at the top of the file, just after any module comments and docstrings,
+/// and before module globals and constants.
+///
+/// ## Example
+/// ```python
+/// def foo():
+///     import os
+///
+///     ...
+/// ```
+///
+/// ## References
+/// - [PEP 8 - Style Guide for Python Code](https://peps.python.org/pep-0008/#imports)
+#[violation]
+pub struct ImportOutsideTopLevel;
+
+impl Violation for ImportOutsideTopLevel {
```

Comment:
Will this rule need special handling to allow typing only imports?

Reply:
I imagine so.  I'll revert this to draft while I sort this out.

Reply:
@MichaReiser I've excluded anything in a type-checking block.  Ideally, we might be able to detect "top-level type-checking blocks" and treat those differently, but I suspect it isn't that important as type-checking blocks are generally at the top-level.